airbyte-commons/src/main/java/io/airbyte/commons/json/JsonPaths.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.commons.json;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.google.api.client.util.Preconditions;
+import com.jayway.jsonpath.Configuration;
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.Option;
+import com.jayway.jsonpath.PathNotFoundException;
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider;
+import com.jayway.jsonpath.spi.json.JsonProvider;
+import com.jayway.jsonpath.spi.mapper.JacksonMappingProvider;
+import com.jayway.jsonpath.spi.mapper.MappingProvider;
+import io.airbyte.commons.util.MoreIterators;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * JSONPath is specification for querying JSON objects. More information about the specification can
+ * be found here: https://goessner.net/articles/JsonPath/. For those familiar with jq, JSONPath will
+ * be most recognizable as "that DSL that jq uses".
+ *
+ * We use a java implementation of this specification (repo: https://github.com/json-path/JsonPath).
+ * This class wraps that implementation to make it easier to leverage this tool internally.
+ *
+ * GOTCHA: Keep in mind with JSONPath, depending on the query, 0, 1, or N values may be returned.
+ * The pattern for handling return values is very much like writing SQL queries. When using it, you
+ * must consider what the number of return values for your query might be. e.g. for this object: {
+ * "alpha": [1, 2, 3] }, this JSONPath "$.alpha[*]", would return: [1, 2, 3], but this one
+ * "$.alpha[0]" would return: [1]. The Java interface we place over this query system defaults to
+ * returning a list for query results. In addition, we provide helper functions that will just
+ * return a single value (see: {@link JsonPaths#getSingleValue(JsonNode, String)}). These should
+ * only be used if it is not possible for a query to return more than one value.
+ */
+public class JsonPaths {
```

Comment:
I could see an argument for keeping this package private. I am not sure if this will mostly be used internally for the `Jsons` and `JsonsSchemas` helpers in the `commons` package versus actually being used in application code. Leaning a little towards starting package private and then opening up if it would be useful.